### PR TITLE
refactor: audit interface

### PR DESCRIPTION
**Pull request by tobiaszheller**
Currently in every external audit implementation both `GetSessionChunk` and `StreamSessionEvents` are either not implemented or are returning nil.

The only one who implements it is main `AuditLog` struct from lib/events package: [GetSessionChunk](https://github.com/gravitational/teleport/blob/a15e987476a0cf78b84aa7751f0b0eee06ff3640/lib/events/auditlog.go#L638-L658) and [StreamSessionEvents](https://github.com/gravitational/teleport/blob/a15e987476a0cf78b84aa7751f0b0eee06ff3640/lib/events/auditlog.go#L923-L1000). Both endpoints are implemented if used with "external log" via downloading recording first.

I have created new interface `ExternalAuditLogger`, which will be embeded into `IAuditLog` and will simplify implementations of external loggers. 

In separate PR I am planning to tackle `GetSessionEvents` methods.

### CODE REVIEW

**Comment by tobiaszheller** [comment]
@zmb3 @espadolini I have added you to review as I have heard that you know that part the best

**Comment by zmb3** [comment]
I haven't looked at the changes yet, but these methods are for session recordings. It would be great if we could start to separate the session recordings and audit log interfaces rather than include all of this functionality in one large interface.

**Comment by tobiaszheller** [comment]
> I haven't looked at the changes yet, but these methods are for session recordings. It would be great if we could start to separate the session recordings and audit log interfaces rather than include all of this functionality in one large interface.

Good idea, I can do that. I just moved it from External Loggers in this PR. I can separate those two either in next PR or change that one, whatever you prefer.

File: lib/events/auditlog.go [hunk 1]
```diff
@@ -1000,7 +1000,7 @@ func (l *AuditLog) StreamSessionEvents(ctx context.Context, sessionID session.ID
 }
 
 // getLocalLog returns the local (file based) audit log.
```
Comment (zmb3) [inline lib/events/auditlog.go:1002]: Is it guaranteed to by a file based audit log still?

**Review by zmb3 — APPROVED** [review]

**Comment by zmb3** [comment]
> I can separate those two either in next PR or change that one, whatever you prefer.

No preference, totally up to you!

File: lib/events/auditlog.go (hunk 1 above)
Comment (tobiaszheller) [reply lib/events/auditlog.go:1002]: do you mean we should update comment that it "returns the local ExternalAuditLogger" or something else?

Comment (zmb3) [reply lib/events/auditlog.go:1002]: Right

**Comment by tobiaszheller** [comment]
> > I can separate those two either in next PR or change that one, whatever you prefer.
> 
> No preference, totally up to you!

I will do it in separate PR then.

**Review by espadolini — APPROVED** [review]
I don't really like the `ExternalAuditLog` name as it seems to imply that it's only implemented by the HA audit logs that store the events somewhere other than the local machine - but I can't really find a better name. Maybe `ActualAuditLog`? 😅

**Comment by espadolini** [comment]
Do we ever need the same object to act as both?

As Go nomenclature goes, the interface that's both `AuditLogger` and `SessionStreamer` should be `AuditLogSessionStreamer`.

**Comment by tobiaszheller** [comment]
> Do we ever need the same object to act as both?

I am afraid that separating this will take a lot time. I can try to tackle it later. 

So what do you think about following:
```
// Replacement of IAuditLog
type AuditLogSessionStreamer interface {
	AuditLogger
	SessionStreamer
}

type SessionStreamer interface {
	GetSessionChunk(namespace string, sid session.ID, offsetBytes, maxBytes int) ([]byte, error)
	StreamSessionEvents(ctx context.Context, sessionID session.ID, startIndex int64) (chan apievents.AuditEvent, chan error)
}

// AuditLogger defines which methods need to implemented by external
// audit loggers.
type AuditLogger interface {
	io.Closer
	EmitAuditEvent(context.Context, apievents.AuditEvent) error
        // GetSessionEvents will be moved to SessionStreamer in separate PR which is already open
	GetSessionEvents(namespace string, sid session.ID, after int, includePrintEvents bool) ([]EventFields, error)
	SearchEvents(fromUTC, toUTC time.Time, namespace string, eventTypes []string, limit int, order types.EventOrder, startKey string) ([]apievents.AuditEvent, string, error)
	SearchSessionEvents(fromUTC, toUTC time.Time, limit int, order types.EventOrder, startKey string, cond *types.WhereExpr, sessionID string) ([]apievents.AuditEvent, string, error)
}
```
is it more readable @zmb3 @espadolini ?

**Comment by espadolini** [comment]
I like it!

Do we need that `io.Closer` in `SessionStreamer` too maybe? And do we even call `Close` on the audit log objects right now? (gravitational/teleport#15210)

**Comment by tobiaszheller** [comment]
@zmb3 Considering this PR and https://github.com/gravitational/teleport/pull/22071, do you think we should backport it? 
I would feel much safer if we just merge it to master and run it via test plan later before release. 

We chatted with Eduardo about it, and there is not active development in audit logs right now, so it might not impact other backports that much.

**Comment by tobiaszheller** [comment]
> Do we need that `io.Closer` in `SessionStreamer` too maybe? And do we even call `Close` on the audit log objects right now? (#15210)

@espadolini I have checked AuditLog struct implementation of SessionStreamer and it seems that Close is not releasing anything from SessionStreamer, it's only in AuditLogger part, so it should be safe to remove it.

From what I have checked, we are not calling that auth anyway, but let's do it in separate PR.

**Comment by tobiaszheller** [comment]
@zmb3 @espadolini I did renaming as agreed in https://github.com/gravitational/teleport/pull/22070#issuecomment-1442287184, PTAL

**Review by zmb3 — APPROVED** [review]

**Review by espadolini — APPROVED** [review]

**Pull request merged**